### PR TITLE
Added Classification and Regression Metrics

### PR DESCRIPTION
**Pull request by jessechancy**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Added regression and classification metrics. Updated LLM Metrics to use new signature.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/metric_definitions.py [hunk 1]
```diff
@@ -16,11 +15,12 @@ def standard_aggregations(scores):
...
+def _validate_text_data(data, metric_name):
+    """Validates that the data is text and is non-empty"""
+    if len(data) == 0:
         return False
 
-    if any(not isinstance(prediction, str) for prediction in predictions):
+    if any(not isinstance(line, str) for line in data):
         _logger.warning(
             f"Cannot calculate {metric_name} for non-string inputs, skipping metric logging."
         )
```
Comment (sunishsheth2009) [inline mlflow/metrics/metric_definitions.py:26]: Can we print the column name and the entire column here so users can debug?

File: mlflow/metrics/metric_definitions.py [hunk 2]
```diff
@@ -151,10 +143,7 @@ def _rouge1_eval_fn(eval_df, metrics):
             )
             return
 
-        y_pred = eval_df["prediction"]
-        predictions = y_pred.squeeze() if isinstance(y_pred, pd.DataFrame) else y_pred
-        references = eval_df["target"]
-
+        references = targets
         scores = rouge.compute(
```
Comment (sunishsheth2009) [inline mlflow/metrics/metric_definitions.py:147]: Can we not do a rename?

```
scores = rouge.compute(
  predictions=predictions,
  references=targets,
  ...
)
```

File: mlflow/metrics/metric_definitions.py [hunk 3]
```diff
@@ -223,8 +216,13 @@ def _rougeL_eval_fn(eval_df, metrics):
         )
 
 
-def _rougeLsum_eval_fn(eval_df, metrics):
-    if "target" in eval_df:
+def _rougeLsum_eval_fn(predictions, targets, metrics):
+    if targets is not None and len(targets) != 0:
+        if not _validate_text_data(targets, "rougeLsum") or not _validate_text_data(
```
Comment (sunishsheth2009) [inline mlflow/metrics/metric_definitions.py:221]: Do we need this `if targets is not None and len(targets) != 0:` check?
Can `_validate_text_data(targets, "rougeLsum")` already handle that?

Comment (sunishsheth2009) [reply mlflow/metrics/metric_definitions.py:221]: Also is there a reason why only some of the metrics have this check to validate if the target has values that are string where other metrics we just check for `if targets is not None and len(targets) != 0:`?

File: mlflow/metrics/metric_definitions.py [hunk 4]
```diff
@@ -249,3 +244,95 @@ def _rougeLsum_eval_fn(eval_df, metrics):
...
             aggregate_results=standard_aggregations(scores),
         )
+
+
+def _mae_eval_fn(predictions, targets, metrics, sample_weight=None):
+    if targets is not None and len(targets) != 0:
+        from sklearn.metrics import mean_absolute_error
+
+        mae = mean_absolute_error(targets, predictions, sample_weight=sample_weight)
+        return MetricValue(aggregate_results={"mean_absolute_error": mae})
```
Comment (sunishsheth2009) [inline mlflow/metrics/metric_definitions.py:254]: What happens in cases where targets is None? Do we want to throw a warning or something that MAE was not computed? Thoughts? 
cc @dbczumar  @annzhang-db 

Note this implies to many of the metrics

File: tests/metrics/test_metric_definitions.py [hunk 5]
```diff
@@ -175,8 +190,68 @@ def test_fails_to_load_metric():
...
+
+def test_r2_score():
+    eval_df = pd.DataFrame({"prediction": [1.0, 2.0, 3.0], "target": [3.0, 2.0, 1.0]})
+    result = r2_score.eval_fn(eval_df["prediction"], eval_df["target"], {})
+    assert result.aggregate_results["r2_score"] == -3.0
+
+
+def test_max_error():
+    eval_df = pd.DataFrame({"prediction": [1.0, 2.0, 3.0], "target": [3.0, 2.0, 1.0]})
+    result = max_error.eval_fn(eval_df["prediction"], eval_df["target"], {})
```
Comment (sunishsheth2009) [inline tests/metrics/test_metric_definitions.py:226]: Nit: Can we refactor these tests to not create eval_df and directly pass predictions and targets to the function?

**Review by sunishsheth2009 — COMMENTED** [review]
Can we also add a test to test this end to end in default-evaluation tests where we can pass some of the metrics as `extra_metrics` and test if the values are computed as expected from `mlflow.evaluate` ?

File: mlflow/metrics/metric_definitions.py (hunk 3 above)
Comment (jessechancy) [reply mlflow/metrics/metric_definitions.py:221]: Some metrics are solely defined for text, while others are not defined for text, and some are defined for both text and numeric values

File: mlflow/metrics/metric_definitions.py (hunk 4 above)
Comment (jessechancy) [reply mlflow/metrics/metric_definitions.py:254]: I believe this issue was discussed on Ann's previous PR too , that's the reason why I kept the behavior like this https://github.com/mlflow/mlflow/pull/9515#discussion_r1318961931

File: tests/evaluate/test_default_evaluator.py [hunk 6]
```diff
@@ -2684,3 +2684,37 @@ def test_eval_df(predictions, targets, metrics, inputs, truth, context):
...
+        model_info = mlflow.pyfunc.log_model(
+            artifact_path="model", python_model=identity_model, input_example=["a", "b"]
+        )
+        data = pd.DataFrame(
+            {
+                "question": ["words random", "This is a sentence."],
+                "answer": ["words random", "This is a sentence."],
+            }
+        )
+        mlflow.evaluate(
```
Comment (sunishsheth2009) [inline tests/evaluate/test_default_evaluator.py:2704]: Can we get the results = mlflow.evaluate

results.metrics has a dict of all the metrics. Can we assert that the metrics in the custom metrics are part of the result dictionary and not ignored

**Review by sunishsheth2009 — APPROVED** [review]
Other than the test comment, it looks good to me.
@dbczumar @annzhang-db any more thoughts here?

**Review by prithvikannan — APPROVED** [review]
LGTM pending Sunish's comments. Awesome work @jessechancy !

**Pull request merged**